packages/block-library/src/navigation/use-block-navigator.js

```diff
@@ -42,6 +42,7 @@ export default function useBlockNavigator( clientId ) {
 			onRequestClose={ () => {
 				setIsNavigationListOpen( false );
 			} }
+			__experimentalPreventKeyboardEventPropagation
```

Comment:
Not a big fan of this prop. Could an alternative be to wrap the modal contents in a container?

```js
<Modal
	...
>
	<div onKeyDown={ ( event ) => {
		// Insert explanatory comment here
		if ( includes( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ], event.keyCode ) ) {
			event.stopPropagation();
		}
	} }>
		<BlockNavigationList ... />
	</div>
</Modal>
```

Alternatively, might be worth looking into if `IgnoreNestedEvents` can help us here. This is how we get `BlockListBlock` to ignore events in other parts of the codebase.

```js
<Modal
	...
>
	<IgnoreNestedEvents childHandledEvents={ [ 'onKeyDown' ] }>
		<BlockNavigationList ... />
	</IgnoreNestedEvents>
</Modal>
```

Reply:
So the issue is that the modal is closed not only when the keyboard nav happens within the modal contents, but also when the modal UI itself is interacted with (e.g. the close button), so surrounding just the BlockNavigationList isn't sufficient.

I've got a different solution which I've split into a separate PR https://github.com/WordPress/gutenberg/pull/19804

I think a better solution is a more global one. I've added it to this PR, but also split it out into #18014